engine/routes.py

```diff
@@ -84,6 +86,114 @@ def decorated_function(*args, **kwargs):
     return inner
 
 
+def set_inactivity_timeout(is_active=lambda _: False, poll_period=None, max_runtime=None):
+    """
+    Decorator to poll whether a function is active and when it becomes inactive, to terminate it. Function activity is
+    checked by an is_active function passed to the decorator. Activity is checked at the end of every poll_period
+    seconds. Also, at the close of each poll period, if the cumulative runtime of the function exceeds max_runtime
+    seconds, the function is terminated. The is_active function has access to poll_period, max_runtime, and all
+    arguments passed to the wrapped function.
+    :param is_active: function used to determine if the decorated function is still active
+                      (e.g. if aggregation is progressing)
+    :param poll_period: the frequency, in seconds, at which the is_active function is called and the decorated
+                        function's runtime is checked against max_runtime
+    :param max_runtime: the maximum amount of time, in seconds, that the decorated function should be allowed to run;
+                        only checked at the end of each poll_period, so ideally max_runtime would be a factor of
```

Comment:
I think you you have the wording backward -- "poll_period should be a factor of max_runtime" or "max_runtime should be a multiple of poll_period"

Reply:
Yeah, "max_runtime should be a multiple of poll_period" is what I want. Will update.